### ISSUE TITLE
SSL tests fail with PR 31

### ISSUE THREAD

**Issue by p-alik** [p-alik/perl-Gearman#32]
PR #31 leads to freeze of [t/18-ssl.t](https://github.com/p-alik/perl-Gearman/blob/master/t/18-ssl.t#L121) in `subtest "sum"` on `do_task`

@esabol got the issue with v2.004.009

see [his comment in #31](https://github.com/p-alik/perl-Gearman/pull/31#issuecomment-346424125)

**Comment by esabol** [comment]
I suspect `Gearman::ResponseParser::parse_sock()` needs to be modified to use `Gearman::Util::_read_sock()`. What do you think @p-alik? Is that possible?

Reference:
http://search.cpan.org/~sullr/IO-Socket-SSL/lib/IO/Socket/SSL.pod#Using_Non-Blocking_Sockets

@p-alik implemented this algorithm in `Gearman::Util::_read_sock()` almost exactly a year ago, but I think we must have overlooked `Gearman::ResponseParser::parse_sock()`. Are there any other sysreads in the code that we missed?

UPDATE: Um, well, I thought this was done, but I see a "TODO" in `Gearman::Util::read_res_packet()` that seemingly indicates it wasn't done...

**Comment by p-alik** [comment]
@esabol, thanks for the hint. Could you test upstream, please.
It seems the commit 527be3d solved the issue.

**Comment by p-alik** [comment]
> Are there any other sysreads in the code that we missed?

no more

**Comment by esabol** [comment]
This is definitely an improvement over 2.004.010, comparable to 2.004.008, but it still freezes up at higher iterations (~1000). Maybe this is a separate issue, but I think the only solution to what I'm seeing is to implement the algorithm at http://search.cpan.org/%7Esullr/IO-Socket-SSL/lib/IO/Socket/SSL.pod#Using_Non-Blocking_Sockets in `Gearman::Util::read_res_packet()`.

**Comment by esabol** [comment]
* worker_ssl.pl:
```
#!/usr/bin/perl
use v5.10;
use strict;
use warnings;

use IO::Socket::IP qw();
use IO::Socket::SSL qw();
use lib '/path/to/github/upstream/perl-Gearman/blib/lib';
use Gearman::Worker;

my $w = Gearman::Worker->new(job_servers => [ {
    'use_ssl'   => 1,
    'ca_file'   => '/path/to/my/server/gearmand-ca.pem',
    'host'      => 'my.gearmand.hostname',
    'port'      => 47301,
    'socket_cb' => sub {
        my $opts = shift;
        $opts->{Timeout} = 60;
        $opts->{SSL_verify_mode} = eval { IO::Socket::SSL::SSL_VERIFY_NONE() };
        return;
    },
} ]);

$w->register_function(
    "foo",
    sub {
        return 'bar';
    },
);

$w->work();
```
* benchmark_ssl_upstream.pl:
```
#!/usr/bin/perl
use v5.10;
use warnings;
use strict;
use Time::HiRes;
use IO::Socket::IP qw();
use IO::Socket::SSL qw();
use Benchmark qw(:hireswallclock);
use Benchmark::Forking qw(cmpthese);
use lib '/path/to/github/upstream/perl-Gearman/blib/lib';
use Gearman::Client;

my ($f, $count) = qw/foo 1000/;
my $c_upstream = Gearman::Client->new(job_servers => [ {
    'use_ssl'   => 1,
    'ca_file'   => '/path/to/my/server/gearmand-ca.pem',
    'host'      => 'my.gearmand.hostname',
    'port'      => 47301,
    'socket_cb' => sub {
        my $opts = shift;
        $opts->{Timeout} = 60;
        $opts->{SSL_verify_mode} = eval { IO::Socket::SSL::SSL_VERIFY_NONE() };
        return;
    },
} ]);

my $r = timethese (
    $count,
    {
        'upstream' => sub { $c_upstream->do_task($f => 'bar',
                                  { 'on_status' => \&update_status }) },
    }
);

#------------------------------------------------------------------------------
sub update_status {

    my($numerator,$denominator) = @_;

    my $percentage = eval { ($numerator / $denominator) * 100.0 };
    return if $@;
    printf("%.0f%%\n",$percentage) if (defined($percentage));

    return;

} #eosub--update_status
#------------------------------------------------------------------------------
```

**Comment by p-alik** [comment]
@esabol, in my environment benchmark stopped at the and with `$count = 1000000`. 

```bash
Benchmark: timing 1000000 iterations of upstream...
  upstream: 601.714 wallclock secs (311.11 usr + 36.66 sys = 347.77 CPU) @ 2875.46/s (n=1000000)
```

**Comment by esabol** [comment]
You are lucky and/or I am unlucky, I guess. Do you have any suggestions on how to debug this? I tried changing DEBUG to 1 in `Gearman::Util`, but it didn't really give me any useful information.

Could be some difference in kernel, networking configuration, or ??? I'm on CentOS 7.4 with 3.10.0-693.5.2.el7.x86_64. We just updated the kernel and all the packages about a week ago, so we should be very current with regards to CentOS. What's yours?

I changed benchmark_upstream_ssl.pl so that I could be sure how many iterations I was actually doing before it would "freeze":
```
my $loop = 0;
my $r = timethese (
    $count,
    {
        'upstream' => sub {
	    $loop++;
	    print STDERR "==> Loop: $loop\n";
	    $c_upstream->do_task($f => 'bar',
				 { 'on_status' => \&update_status });
	},
    }
);
```
Then, I did some experimenting with the size of the string returned by the worker and adding a sleep of 0.5 seconds to the worker. If I increased the size of the string being returned from 3 characters to 300, I could routinely get 2000+ iterations. Adding the sleep sometimes improved things when the string was only 3 characters, but it was inconsistent. If I increased the size of the string to 600 characters, it got worse, not better. It's all fairly inconsistent voodoo magic though.

I am starting 5 workers. How many are you starting? Hmmm, I tried with only one worker, and that seemed to improve it actually. I could routinely (but not always) get 3000+ iterations.

I still think the implementation does not handle either the various SSL connection states or renegotiation correctly. Do you feel otherwise? I don't see any handling of EWOULDBLOCK or SSL_CAN_READ or SSL_CAN_WRITE states. I have tried to implement the while loop from http://search.cpan.org/%7Esullr/IO-Socket-SSL/lib/IO/Socket/SSL.pod#Using_Non-Blocking_Sockets but I can't seem to get it working with the other code in `read_res_packet()`, starting at `if (!defined $type) {` and the rest. I'm not seeing how to incorporate that into the example while loop. I'm not very good at event loop programming, I think.

**Comment by esabol** [comment]
Setting DEBUG to 1 in Gearman::Util, I typically see the following when a "freeze" happens (on the 127th iteration in this case):
```
==> Loop: 127
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
  We have a timeout of 30
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
 Fully formed res packet, returning; type=noop len=0
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
   Finished reading
 Fully formed res packet, returning; type=job_assign len=28
   Entering read loop
   Finished reading
   Finished reading
 Fully formed res packet, returning; type=job_created len=20
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
 Fully formed res packet, returning; type=no_job len=0
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
 Fully formed res packet, returning; type=no_job len=0
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
 Fully formed res packet, returning; type=no_job len=0
 Entering read_res_packet at /path/to/github/upstream/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
 Fully formed res packet, returning; type=no_job len=0
```
And then the last 5 lines just keep repeating every few seconds.

At that point, gearmand status shows no jobs queued or running. I guess the client is hanging on reading the result from gearmand?

**Comment by esabol** [comment]
I added a bunch of `warn`s to `Gearman::Client::do_task()`, and it's now clear to me that the client is hanging at `$ts->wait(timeout => $task->timeout);` in `do_task()`.

**Comment by esabol** [comment]
After adding some warns to `Gearman::Taskset::wait()`...
```
Client: after add_task, waiting (timeout=undef)...
Taskset::wait: before while loop
Taskset::wait: before select
 Entering read_res_packet at /path/to/github/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
   Entering read loop
   Finished reading
 Fully formed res packet, returning; type=no_job len=0
Client ==>
Taskset::wait: after select, nfound = undef
Taskset::wait: before select
Taskset::wait: after select, nfound = undef
Taskset::wait: before select
Taskset::wait: after select, nfound = undef
...
```
The last two lines just keep repeating...

**Comment by esabol** [comment]
Potentially relevant articles:
http://www.perlmonks.org/?node_id=451222
http://www.perlmonks.org/?node_id=845342

Except $IO::Socket::SSL_ERROR and $! are both undef before and after the select call, so I'm a bit lost.

**Comment by esabol** [comment]
OK, I think I see the problem now. In `Gearman::Util::_read_sock()`, whenever `$sock->pending()` is non-zero after the `sysread()` there, the `select` in `Taskset::wait()` never returns a defined value.
```
Client: after new_task_set
 Entering read_res_packet at /Heasarc_dev/heasarc/hera/lib/github/perl-Gearman/blib/lib/Gearman/Util.pm line 113.
 Starting up event loop
  We have a timeout of 30
   Entering read loop
_read_sock (before sysread): pending(sock) = undef
_read_sock (after sysread): rv = 12, readlen = 12, pending(sock) = 52, SSL_ERROR = undef
   Pending(full) bytes to read 52 bytes
   Finished reading
_read_sock (before sysread): pending(sock) = 52
_read_sock (after sysread): rv = 18, readlen = 18, pending(sock) = 34, SSL_ERROR = undef
   Pending(full) bytes to read 34 bytes
   Finished reading
 Fully formed res packet, returning; type=job_created len=18
Client: after add_task, waiting (timeout=undef)...
Taskset::wait: before while loop, $IO::Socket::SSL_ERROR = undef;
Taskset::wait: before select, time_left = 0.5
Taskset::wait: after select, $! = undef, nfound = undef, $IO::Socket::SSL_ERROR = undef;
Taskset::wait: before select, time_left = 0.5
Taskset::wait: after select, $! = undef, nfound = undef, $IO::Socket::SSL_ERROR = undef;
```

**Comment by esabol** [comment]
The IO::Socket::SSL pod says:
> Thus, in order to decide if you can read more data (e.g. if sysread will block) you must check if there are still data in the current SSL frame by calling pending and if there are no data pending you might check the underlying socket with select or poll. Another way might be if you try to sysread at least 16kByte all the time. 16kByte is the maximum size of an SSL frame and because sysread returns data from only a single SSL frame you can guarantee that there are no pending data.

So I thought maybe reading the pending bytes would clear the buffer and allow the select in Taskset::wait to work. So I added:
```Perl
	if (my $pending = $sock->pending()) {
	    warn " Pending(read_res_packet) bytes to read $pending bytes\n" if DEBUG;
	    my $bufp = '';
	    my $retp = sysread($sock, $bufp, $pending);
	    warn " Pending bufp = \"\n$bufp\"\n" if DEBUG;
	}
```
to `Gearman::Util::read_res_packet()` after the `warn " Fully formed res packet...` line.

That resulted in the following being printed:
```
 Fully formed res packet, returning; type=job_created len=18
 Pending(read_res_packet) bytes to read 34 bytes
 Pending bufp = "
H:gearmandev:48597bar"
```
That appears to be the result that `Taskset::wait()` is waiting on? So I think `Taskset::wait()` needs to check `$sock->pending` before calling `select` or something like that. And if pending, then skip the select.

**Comment by esabol** [comment]
> So I think Taskset::wait() needs to check $sock->pending before calling select or something like that. And if pending, then skip the select.

Well, I tried to implement that, but it's still not working. In that scenario, it just freezes after `Starting up event loop`... :( Well, my work day is done here, so I'm going to give it a rest until Monday. Let me know if you have any insights....

**Comment by p-alik** [comment]
> Could be some difference in kernel, networking configuration, or ??? I'm on CentOS 7.4 with 3.10.0-693.5.2.el7.x86_64. We just updated the kernel and all the packages about a week ago, so we should be very current with regards to CentOS. What's yours?

I'm using Ubuntu 16.04 right now with 4.10.0-40-generic kernel. Local networking.

> After adding some warns to Gearman::Taskset::wait()...

See [this comment at #31 ](https://github.com/p-alik/perl-Gearman/pull/31#issuecomment-346115363)

I'll try to setup CentOS 7.4 environment fro testing purpose.

**Comment by esabol** [comment]
Based on what I've read, it could be due to a difference in the SSL cipher being used for the connection as well.

Yesterday, I added checks for `$sock->pending()` to both `Taskset::wait()` and `Util::read_res_packet()`, and that seemed to resolve the issue I was seeing. I need to do more testing, and I need to clean up the changes quite a bit, but I hope to post some diffs soon.

**Comment by esabol** [comment]
Here is a patch for Taskset.pm and Util.pm which adds checks for pending bytes prior to select. It could probably be tidied up quite a bit more in order to have less duplicate code (especially in `Taskset::wait`). I added some `warn`s. Feel free to tweak as you see fit. What do you think?

[diff.txt](https://github.com/p-alik/perl-Gearman/files/1533100/diff.txt)

**Comment by p-alik** [comment]
Thank you, @esabol. I could reproduce the behavior you described in my CentOS vagrant box.

**Comment by esabol** [comment]
Glad to hear you can reproduce it, @p-alik.

**Comment by p-alik** [comment]
@esabol, my testing based on upstream 0a723e2 doesn't fail anymore. `Taskset->wait` doen't stuck anymore on `select` by using SSL connection to only with one ` gearmand`.

Testing with multiple `gearmand` instances is the next step.

I thank you once again for your contribution!

**Comment by esabol** [comment]
Comments added on upstream commits. Thanks for everything, @p-alik!

**Comment by esabol** [comment]
I've tested the latest upstream branch many times and have encountered no problems. My timings of 10000 iterations can vary quite a bit though, ranging from 13 seconds to 32 seconds. I cannot explain such a wide disparity in timings. Anecdotal individual response times appear to be comparable to or perhaps even faster than 2.004.008, so I'm happy about that.

**Comment by p-alik** [comment]
> My timings of 10000 iterations can vary quite a bit though, ranging from 13 seconds to 32 seconds. I cannot explain such a wide disparity in timings.

There is no such timing disparity in my local environment. Here some benchmark runs on upstream: 
```bash
$ for i in `seq 1 10`; do perl -Ilib tests/issue32-client-benchmark.pl; done
Benchmark: timing 10000 iterations of upstream...
  upstream: 8.16361 wallclock secs ( 3.50 usr +  0.47 sys =  3.97 CPU) @ 2518.89/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.94602 wallclock secs ( 3.62 usr +  0.40 sys =  4.02 CPU) @ 2487.56/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.34276 wallclock secs ( 3.57 usr +  0.32 sys =  3.89 CPU) @ 2570.69/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.38601 wallclock secs ( 3.60 usr +  0.32 sys =  3.92 CPU) @ 2551.02/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.68615 wallclock secs ( 3.55 usr +  0.40 sys =  3.95 CPU) @ 2531.65/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 8.09272 wallclock secs ( 3.62 usr +  0.37 sys =  3.99 CPU) @ 2506.27/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.32726 wallclock secs ( 3.57 usr +  0.34 sys =  3.91 CPU) @ 2557.54/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 8.20304 wallclock secs ( 3.59 usr +  0.43 sys =  4.02 CPU) @ 2487.56/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.38598 wallclock secs ( 3.54 usr +  0.40 sys =  3.94 CPU) @ 2538.07/s (n=10000)
Benchmark: timing 10000 iterations of upstream...
  upstream: 7.54255 wallclock secs ( 3.65 usr +  0.30 sys =  3.95 CPU) @ 2531.65/s (n=10000)
```
v2.004.008 timings are similar.

I haven't finished testing of upstream yet, but I'll do it this week.

**Comment by p-alik** [comment]
benchmark runs on v2.004.008 freeze sporadically. Tested on Ubuntu 16.04.

**Comment by esabol** [comment]
As long as it doesn't freeze with upstream, that's fine. :)

**Comment by p-alik** [comment]
v2.004.011 is released to CPAN. Thank you, @esabol.
